Sources/SwiftSyntax/SyntaxRewriter.swift.gyb

```diff
@@ -111,198 +111,177 @@ public enum SyntaxVisitorContinueKind {
   case skipChildren
 }
 
-open class SyntaxVisitor {
-    public init() {}
+public protocol SyntaxVisitor {
 % for node in SYNTAX_NODES:
 %   if is_visitable(node):
-  /// Visting ${node.name} specifically.
+  /// Visiting `${node.name}` specifically.
   ///   - Parameter node: the node we are visiting.
   ///   - Returns: how should we continue visiting.
-  open func visit(_ node: ${node.name}) -> SyntaxVisitorContinueKind {
-    return .visitChildren
-  }
+  mutating func visit(_ node: ${node.name}) -> SyntaxVisitorContinueKind
```

Comment:
It's worth documenting why these functions are mutating because from the clients' perspective, visitation should be side-effect free. Any reason we specify `mutating` here rather than delegating to conformers to decide?
This means if they want mutating version, clients can use `class` type; otherwise `struct` type.

Reply:
> visitation should be side-effect free

That is dependent on the specific visitor which may not be side-effect free. For example let's say you write a visitor to count the number of functions by increasing a counter member, if you make it a `struct` its visitor method need to be `mutating`. Also see the implementation of `SyntaxVerifier`.

Reply:
Sure, there are cases that visitor needs to be mutating. I guess my main concern is when we put such keyword in the protocol, it looks as if a visitor is required to be mutating.

Reply:
A mutating visitor is the common case, there's not much useful things you can do with a non-mutating visitor.
I don't think we need to put documentation in every `mutating` function to explain in case someone doesn't understand what `mutating` means in a protocol, the StdLib does not do that.

Reply:
Clients can still implement mutating visitor if we drop the mutating keyword in the protocol; they just need to implement the conforming type by using a class.

Having mutating keyword in the protocol may make nonmutating conformers look like an anti-patter, IMO. I think whether a visitor function mutating or not is domain-specific, thus it doesn’t belong to the protocol.

Reply:
> Clients can still implement mutating visitor if we drop the mutating keyword in the protocol; they just need to implement the conforming type by using a class.

What would be the benefit of this, that would practically exclude value types from conforming to the protocol. You don't have to put the keyword if the conforming type is a class, this is relevant mainly for value types.

> I think whether a visitor function mutating or not is domain-specific, thus it doesn’t belong to the protocol.

`mutating` in the protocol is important for value types to participate in protocol-based programming, I would recommend to review how it is used in the StdLib, for example in `TextOutputStream` among other cases.

Reply:
`mutating` means you *allow* value types to mutate state during visitation, it's not a *requirement*.

Reply:
That’s fair. Value type visitor will be happier with this change I presume.